src/analysis/processing/qgsalgorithmvoronoipolygons.cpp

```diff
@@ -0,0 +1,194 @@
+{
+  return QObject::tr( "Generates a polygon layer containing the Voronoi diagram corresponding to input points." );
+}
+
+QgsVoronoiPolygonsAlgorithm *QgsVoronoiPolygonsAlgorithm::createInstance() const
+{
+  return new QgsVoronoiPolygonsAlgorithm();
+}
+
+void QgsVoronoiPolygonsAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ), QList< int >() << QgsProcessing::TypeVectorPoint ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "BUFFER" ), QObject::tr( "Buffer region (% of extent)" ), QgsProcessingParameterNumber::Double, 0, false, 0 ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "TOLERANCE" ), QObject::tr( "Tolerance" ), QgsProcessingParameterNumber::Double, 0, true, 0 ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "COPY_ATTRIBUTES" ), QObject::tr( "Copy attributes from input features" ), true ) );
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Voronoi polygons" ), QgsProcessing::TypeVectorPolygon ) );
+}
+
+QVariantMap QgsVoronoiPolygonsAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  std::unique_ptr< QgsProcessingFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  if ( source->featureCount() < 3 )
+    throw QgsProcessingException( QObject::tr( "Input layer should contain at least 3 points." ) );
+
+  const double buffer = parameterAsDouble( parameters, QStringLiteral( "BUFFER" ), context );
+  const double tolerance = parameterAsDouble( parameters, QStringLiteral( "TOLERANCE" ), context );
+  const bool copyAttributes = parameterAsBool( parameters, QStringLiteral( "COPY_ATTRIBUTES" ), context );
+
+  QgsFields fields;
+  if ( copyAttributes )
+  {
+    fields = source->fields();
+  }
+  else
+  {
+    fields.append( QgsField( QStringLiteral( "id" ), QVariant::LongLong ) );
+  }
+
+  QString dest;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, dest, fields, Qgis::WkbType::Polygon, source->sourceCrs() ) );
+  if ( !sink )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  QgsFeatureRequest request;
+  if ( !copyAttributes )
+  {
+    request.setSubsetOfAttributes( QList< int >() );
+  }
+  QgsFeatureIterator it = source->getFeatures( request, QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks );
+
+  QgsGeometry allPoints;
+  QHash< QgsFeatureId, QgsAttributes > attributeCache;
+
+  long i = 0;
+  double step = source->featureCount() > 0 ? 50.0 / source->featureCount() : 1;
+
+  const QgsSpatialIndex index( it, [&]( const QgsFeature & f )->bool
```

Comment:
This only needs to happen if copyAttributes is true. Otherwise we can skip the cost of creating the index.

And if copyAttributes is true then we could potentially use https://github.com/libgeos/geos/pull/781 to avoid the index for newer geos versions